owncloudData/src/main/java/com/owncloud/android/data/authentication/datasources/implementation/OCLocalAuthenticationDataSource.kt

```diff
@@ -0,0 +1,245 @@
+        userName: String,
+        password: String,
+        serverInfo: ServerInfo,
+        userInfo: UserInfo?,
+        updateIfAlreadyExists: Boolean
+    ): String =
+        addAccount(
+            lastPermanentLocation,
+            serverInfo,
+            userName,
+            password,
+            updateIfAlreadyExists
+        ).also {
+            updateUserAndServerInfo(it, serverInfo, userInfo, userName)
+        }.name
+
+    override fun addOAuthAccount(
+        lastPermanentLocation: String?,
+        userName: String,
+        authTokenType: String,
+        accessToken: String,
+        serverInfo: ServerInfo,
+        userInfo: UserInfo?,
+        refreshToken: String,
+        scope: String?,
+        updateIfAlreadyExists: Boolean
+    ): String =
+        addAccount(
+            lastPermanentLocation,
+            serverInfo,
+            userName,
+            updateIfAlreadyExists = updateIfAlreadyExists
+        ).also {
+            updateUserAndServerInfo(it, serverInfo, userInfo, userName)
+
+            accountManager.setAuthToken(it, authTokenType, accessToken)
+
+            accountManager.setUserData(it, KEY_SUPPORTS_OAUTH2, "TRUE")
+            accountManager.setUserData(it, KEY_OAUTH2_REFRESH_TOKEN, refreshToken)
+            scope?.run {
+                accountManager.setUserData(it, KEY_OAUTH2_SCOPE, this)
+            }
+        }.name
+
+    /**
+     * Add new account to account manager, shared preferences and returns account name
+     */
+    private fun addAccount(
+        lastPermanentLocation: String?,
+        serverInfo: ServerInfo,
+        userName: String,
+        password: String = "",
+        updateIfAlreadyExists: Boolean = false
+    ): Account {
+        if (lastPermanentLocation != null) {
+            serverInfo.baseUrl = WebdavUtils.trimWebdavSuffix(lastPermanentLocation)
+        }
+
+        val uri = Uri.parse(serverInfo.baseUrl)
+
+        val accountName = AccountUtils.buildAccountName(uri, userName)
+
+        if (accountExists(accountName)) {
+            if (!updateIfAlreadyExists) {
+                Timber.d("The account already exists")
+                throw AccountNotNewException()
+            } else {
+                return getAccountByName(accountName)!! // Account won't be null, since we've already checked it exists
+            }
+        } else {
+            val newAccount = Account(accountName, accountType)
+
+            // with external authorizations, the password is never input in the app
+            accountManager.addAccountExplicitly(newAccount, password, null)
+
+            /// add the new account as default in preferences, if there is none already
+            val defaultAccount: Account? = getCurrentAccount()
+            if (defaultAccount == null) {
+                val editor = PreferenceManager
+                    .getDefaultSharedPreferences(context).edit()
+                editor.putString("select_oc_account", accountName)
```

Comment:
I'd rather create an `AuthenticationConstants` file in the data layer so the constant is independent of the implementation

Reply:
Sure, i meant to create a  `SharedPreferencesConstants` to include there every constant related with Shared Preferences.

Reply:
I understand what you meant but I included the constant in a `AuthenticationConstants` file since we do not need to know if we save the selected account in the SharedPreferences or whatever we use under the hood